src/wacksbywarby/etsy.py

```diff
@@ -1,40 +1,190 @@
 import logging
 import os
-from typing import Dict
+import secrets
+import hashlib
+import base64
+import uuid
+from datetime import datetime
 
 import requests
-
-from wacksbywarby.models import Inventory
+from dotenv import load_dotenv
+from wacksbywarby.models import Sale
+from wacksbywarby.db import Wackabase
 
 logger = logging.getLogger("etsy")
 
 
 class Etsy:
-    def __init__(self, debug=False) -> None:
-        self.key = os.getenv("ETSY_API_KEY")
-        self.shop_name = "wicksbywerby"
+    def __init__(self, credentials=None, debug=False) -> None:
         self.debug = debug
+        self.shop_id = os.getenv("ETSY_SHOP_ID")
+        self.v3_api_key = os.getenv("ETSY_V3_API_KEY")
+
+        v3_access_token = credentials.access_token
+        self.credentials = credentials
+        self.headers = {
+            'x-api-key': self.v3_api_key,
+            'Authorization': f'Bearer {v3_access_token}'
+        }
+
+        # only need when getting initial access token for v3
+        self.challenge_verifier = os.getenv('ETSY_CHALLENGE_VERIFIER')
+        self.redirect_uri = os.getenv('ETSY_REDIRECT_URI')
+
+    def _generate_challenge_verifier(self):
+        # generate PKCE challenger verifier for etsy oauth which can be stored in env and re-used
+        token_length = 44
+        code_verifier = secrets.token_urlsafe(32)[:token_length]
+        hashed = hashlib.sha256(code_verifier.encode('ascii')).digest()
+        encoded = base64.urlsafe_b64encode(hashed)
+        code_challenge = encoded.decode('ascii')[:-1]
+        print('verifier', code_verifier, 'challenge', code_challenge)
+
+    def make_oath_connect_request(self):
+        """
+        Make initial request to prompt Etsy account owner to grant permissions to our Etsy app.
+        https://developers.etsy.com/documentation/essentials/authentication#step-1-request-an-authorization-code
+        """
+        connect_url = "https://www.etsy.com/oauth/connect"
+        state = uuid.uuid4().hex
+        hashed = hashlib.sha256(self.challenge_verifier.encode('ascii')).digest()
+        encoded = base64.urlsafe_b64encode(hashed)
+        code_challenge = encoded.decode('ascii')[:-1]
+        # A URL-encoded, space-separated list of one or more scopes (e.g., shops_r%20shops_w)
+        scopes = ' '.join(['listings_r', 'transactions_r'])
+        # scopes = ' '.join(['transactions_r'])
```

Comment:
```suggestion
```

Reply:
the things i'm less confident about are 1) the auth token is expired by now so we'll need to rely on the refresh cron to refresh that before any requests will work unless we refresh and update them manually 2) i had updated number of sales and last sale date manually. wack will theoretically start from that point and pick up all the sales until now. hopefully isn't too too many sales or something might break